Rough Idea: Replace NETWORK=testnet with `HttpBridge 'Testnet`

Anviking opened the pull request:
Inspired by #313 

This way, instead of having a `HttpBridge` target, which secretly have three supported networks in the environment variable, we get `HttpBridge 'Mainnet`, `HttpBridge 'Testnet` and `HttpBridge 'Staging` as different backends.

Example where it plays out nicely:
```haskell
instance KnownNetwork n => KeyToAddress (HttpBridge n) where
    keyToAddress key =
        Address
            $ CBOR.toStrictByteString
            $ CBOR.encodeAddress xpub encodeAttributes
      where
        xpub = getKey key
        encodeAttributes = case demoteNetwork (Proxy @n) of
            Mainnet -> emptyAttributes
            Staging -> emptyAttributes
            Testnet -> attributesWithProtocolMagic (protocolMagic network)
```

_should have been a draft pr_

piotr-iohk commented on test/integration/Test/Integration/Scenario/EosWallets.hs:
```diff
@@ -50,14 +50,88 @@ spec = do
+            verify (response :: Either ClientError EosWallet)
+                [ expectJSONError $ case poolGap of
+                                        -1 -> "Error in $.addressPoolGap: Word8 is either floating or will cause over or underflow: -1.0"
+                                        _  -> "Error in $.addressPoolGap: Address pool gap should be in range [10..100]"
```
I like this error message `Address pool gap should be in range [10..100]`. I was expecting it also when providing `-1` as `addressPoolGap`. Instead we have `Word8 is either floating or will cause over or underflow: -1.0`. Maybe we'd like to change it for consistency?

akegalj replied:
> Error in $.addressPoolGap: Word8 is either floating or will cause over or underflow: -1.0

msg is interesting one - I didn't expect to see something like this. For example when I try to manually create address pool gap with a negative number:
```haskell
> AddressPoolGap (-1)
```
I get
```haskell
    Literal -1 is out of the Word8 range 0..255
```
which makes sense as Word8 is defined in specified range.

Edit:
Aha - msg comes from `Aeson` and decoding a "-1" string into a `Word8`:

```haskell
> import Data.Aeson
> eitherDecode "-1" :: Either String Word8
Left "Error in $: Word8 is either floating or will cause over or underflow: -1.0"
```

so everything is as expected here. So you will try to push your json to the api. First thing that api will do is it will try to decode this json into expected Haskell data type. This will immediatelly fail in this json decoding part as Word8 won't be able to decode "-1" and `Aeson` library will rise the exception. If the exception wasn't rased by underlying aeson lib, we would catch it and rise "Address pool gap should be in range [10..100]". 

If we would like to rise only single exception for `AddressPoolGap` one could define custom `FromJSON` instance that works only for `[10..100]` (ie, `[minBound..maxBound]`) and not to use custom/generic aeson instance https://github.com/input-output-hk/cardano-wallet/blob/15dd91b05112c16be506129e5c00e88114f8bd16/src/Cardano/Wallet/Kernel/AddressPoolGap.hs#L33 . This would do the trick in our case:
```haskell
instance FromJSON AddressPoolGap where
    parseJSON = parseJSON >=> \gap -> case mkAddressPoolGap gap of
        Left err -> fail $ toString $ sformat build $ bprint
        ("Address pool gap should be in range ["%int%".."%int%"].")
        (minBound @AddressPoolGap)
        (maxBound @AddressPoolGap)
        Right x  -> pure x
```

or something along these lines. Anyway if you think this would be nice to have please open a new ticket and we will discuss it with rest of the team. For now I don't think this is necessary nor urgent to have

piotr-iohk commented:
@denisshevchenko - I'll ask here as it is related to creating EOS wallets. In the API doc -> https://input-output-hk.github.io/cardano-wallet/#tag/Externally-Owned-Wallets%2Fpaths%2F~1api~1v1~1wallets~1externally-owned%2Fpost it is said `Creates a new or restores an existing Wallet`. Is there an option to `restore` or the mention about the `restore` should be actually removed from the doc?

denisshevchenko commented:
> Is there an option to `restore`

@piotr-iohk Not yet.

akegalj commented on test/integration/Test/Integration/Scenario/EosWallets.hs:
```diff
@@ -48,16 +49,521 @@ spec = do
+            "name": "My EOS Wallet"
+            }|]
+        verify (response :: Either ClientError EosWallet)
+            [ expectJSONError "Error in $.accounts[0]: When parsing the record accountPublicKeyWithIx of type Cardano.Wallet.API.V1.Types.AccountPublicKeyWithIx the key publicKey was not present."
```
hm - these types of error checks will gratelly depend on aeson version (msg might be modified in aeson) and on naming of the module where the function is located (if someone changes name of the module or name of the function itself msg will change)

I guess its ok as a temp solution. What do you think @denisshevchenko @uroboros ?

denisshevchenko replied:
Agree, phrase in `expectJSONError` is too long and specific.

piotr-iohk replied:
As for the error messages `Address pool gap should be in range [10..100]` and `Word8 is either floating or will cause over or underflow` I'd be for having one consistent message and obviously the first one is more user friendly and informative. I'll create a ticket. This is actually observed for `addressPoolGap` and `index` fields for certain values outside their ranges...

>depend on ... naming of the module where the function is located (if someone changes name of the module or name of the function itself msg will change)

makes sense, I'll shorten the messages. :+1:

akegalj commented on test/integration/Test/Integration/Scenario/EosWallets.hs:
```diff
@@ -48,16 +49,521 @@ spec = do
+                "name": "My EosWallet2"
+                }|]
+        verify (updResp :: Either ClientError EosWallet)
+            [ expectWalletError (UnknownError "UpdateEosWalletError")
```
I don't think this is right error - right @denisshevchenko @uroboros ?

I believe we are missing case such as `<|> try' @UpdateEosWalletError updateEosWalletError` at https://github.com/input-output-hk/cardano-wallet/blob/develop/src/Cardano/Wallet/API/V1/ReifyWalletError.hs#L47

akegalj replied:
fixed in my local branch - will push after we merge #337 (which fixes some failures) and test the patch

piotr-iohk commented:
@akegalj I have created a ticket #342 and marked test `EOSWALLETS_LIST_01, EOSWALLETS_LIST_02 - cannot get FO wallets with EOS endpoint` with `pendingWith`. From my side it is ready to be merged.

Anviking committed:
Move 'network' at the type-level, as part of the target type

KtorZ commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge/Environment.hs:
```diff
@@ -41,6 +44,10 @@ import qualified Data.Text as T
 
+
+class KnownNetwork n where
+    demoteNetwork :: Proxy n -> Network
```
We'll probably need a type family here since we have different network between the bridge and jormungandr (and maybe with haskell shelley too)..  So, something like:

```hs
class KnownNetwork n where
    type Network n :: *
    demoteNetwork :: Proxy n -> Network n
```

This leave free the `Network n` type to be implemented by whoever implement the instance.

Anviking replied:
I'd be more interested in assuming nothing about `target` from `cardano-wallet-core`, and maybe not even having `KnownNetwork` anywhere. I think that might be possible 🤔

KtorZ reviewed (commented):
I like the idea and it's actually quite much safer than having a global ENV var for this :thinking: ...

KtorZ committed:
resolve remaining type problem. Network is now fully embedded in the 'HttpBridge' target

KtorZ committed:
remove 'Cardano.Wallet.Environment' and unit test spec indirections

KtorZ committed:
remove now useless travis job stage

We are indeed now able to perform tests against both mainnet and testnet at the same type
via some type trickery. No need for an extra ENV var anymore.

KtorZ committed:
remove redundant dependencies from jormungandr .cabal

Anviking commented on lib/http-bridge/test/unit/Cardano/Wallet/HttpBridge/BinarySpec.hs:
```diff
@@ -58,6 +62,9 @@ import qualified Data.ByteString.Lazy as BL
 
 {-# ANN spec ("HLint: ignore Use head" :: String) #-}
+
+type N = 'Testnet
```
@KtorZ I added it here when I was experimenting, but I take it that it should be here anymore.

KtorZ replied:
Indeed. Type-aliases (especially this one) usually provides very little value. They make deciphering of code harder and have a tendency to spread.

KtorZ committed:
Adjust remaining spec files in 'http-bridge'

Relevant specs now run for both mainnet and testnet.
